File: week3/1260/1260.py
from queue import Queue

class Node:
    def __init__(self, num):
        self.num = num
        self.connect = []
        self.visit = False

    def connect_node(self, node_num):
        self.connect.append(node_num)

class P_1260:
    def __init__(self):
        self.N, self.M, self.V = map(int, input().split())
        self.nodes = {}
        for _ in range(self.M):
            node1, node2 = map(int, input().split())
            if node1 not in self.nodes:
                self.nodes[node1] = Node(node1)
            if node2 not in self.nodes:
                self.nodes[node2] = Node(node2)
            self.nodes[node1].connect_node(node2)
            self.nodes[node2].connect_node(node1)
        for _, node in self.nodes.items():
            node.connect.sort()
        # no connected nodes
        if self.V not in self.nodes:
            self.nodes[self.V] = Node(self.V)

    def run(self):
        start_node = self.nodes[self.V]
        self.dfs_recursive(start_node)
        print()
        self.reset_visit()
        # self.dfs_stack(start_node)
        # print()
        # self.reset_visit()
        self.bfs(start_node)

    def dfs_recursive(self, node):
        print(node.num, end=' ')
        node.visit = True
        
        for next_node_num in node.connect:
            next_node = self.nodes[next_node_num]
            if next_node.visit:
                continue
            self.dfs_recursive(self.nodes[next_node_num])

    def dfs_stack(self, node):
        stack = [node]

        while len(stack) > 0:
            top = stack.pop()
            if top.visit:
                continue

            print(top.num, end=' ')
            top.visit = True # visit check when printing
            
            for next_node_num in reversed(top.connect):
                next_node = self.nodes[next_node_num]
                if next_node.visit:
                    continue
                stack.append(next_node)

    def bfs(self, node):
        queue = Queue()
        queue.put(node)
        node.visit = True
        
        while not queue.empty():
            front = queue.get()
            print(front.num, end= ' ')

            for next_node_num in front.connect:
                next_node = self.nodes[next_node_num]
                if next_node.visit:
                    continue
                next_node.visit = True
                queue.put(next_node)

    def reset_visit(self):
        for _, node in self.nodes.items():
            node.visit = False


if __name__ == "__main__":
    P = P_1260()
    P.run()

File: with_friends/week3/1260/1260.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <stack>
#include <queue>

using namespace std;

struct Node {
    int num;
    vector<int> connected_num;
    bool visit;
};

int N, M, V;

unordered_map<int, Node> nodes;

void set_input() {
    cin >> N >> M >> V;

    int node_num1, node_num2;
    for (int i=0; i<M; i++) {
        cin >> node_num1 >> node_num2;
        if (nodes.find(node_num1) == nodes.end()) {
            Node node;
            node.num = node_num1;
            node.visit = false;
            nodes.insert(make_pair(node_num1, node));
        }
        if (nodes.find(node_num2) == nodes.end()) {
            Node node;
            node.num = node_num2;
            node.visit = false;
            nodes.insert(make_pair(node_num2, node));
        }
        nodes.find(node_num1)->second.connected_num.push_back(node_num2);
        nodes.find(node_num2)->second.connected_num.push_back(node_num1);
    }
    // sort
    for (auto it=nodes.begin(); it!=nodes.end(); it++)
        sort(it->second.connected_num.begin(), it->second.connected_num.end());
    // no connected node from V
    if (nodes.find(V) == nodes.end()) {
        Node node;
        node.num = V;
        node.visit = false;
        nodes.insert(make_pair(V, node));
    }
}

void dfs_recursive(Node* node) {
    cout << node->num << ' ';
    node->visit = true;

    for (auto node_num : node->connected_num) {
        Node* next_node = &(nodes.find(node_num)->second);
        if (next_node->visit == true)
            continue;
        dfs_recursive(next_node);
    }
}

void dfs_stack() {
    stack<Node*> s;
    s.push(&(nodes.find(V)->second));
    
    while (!s.empty()) {
        Node* now = s.top();
        s.pop();
        if (now->visit)
            continue;
        
        cout << now->num << ' ';
        now->visit = true;

        for (auto it=now->connected_num.rbegin(); it!=now->connected_num.rend(); it++) {
            Node* next = &(nodes.find(*it)->second);
            if (nodes.find(*it)->second.visit)
                continue;
            s.push(next);
        }
    }
}

void bfs() {
    queue<Node*> q;
    q.push(&(nodes.find(V)->second));
    nodes.find(V)->second.visit = true;

    while (!q.empty()) {
        Node* now = q.front();
        q.pop();

        cout << now->num << ' ';
        
        for (auto it=now->connected_num.begin(); it!=now->connected_num.end(); it++) {
            Node* next = &(nodes.find(*it)->second);
            if (next->visit)
                continue;
            q.push(next);
            next->visit = true;
        }
    }
}

void reset_visit() {
    for (auto it=nodes.begin(); it!=nodes.end(); it++)
        it->second.visit = false;
}

int main() {
    set_input();
    // dfs_recursive(&nodes.find(V)->second);
    // cout << endl;
    dfs_stack();
    cout << endl;
    reset_visit();
    bfs();
}